// @percy/cli-snapshot/dist/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.migration = migration;
exports.schema = void 0;
const schema = {
  static: {
    type: 'object',
    additionalProperties: false,
    properties: {
      baseUrl: {
        type: 'string',
        pattern: '^/',
        default: '/',
        errors: {
          pattern: 'must start with a forward slash (/)'
        }
      },
      files: {
        anyOf: [{
          type: 'string'
        }, {
          type: 'array',
          items: {
            type: 'string'
          }
        }],
        default: '**/*.{html,htm}'
      },
      ignore: {
        anyOf: [{
          type: 'string'
        }, {
          type: 'array',
          items: {
            type: 'string'
          }
        }],
        default: ''
      }
    }
  }
};
exports.schema = schema;

function migration(config, {
  map,
  del
}) {
  /* eslint-disable curly */
  if (config.version < 2) {
    // static-snapshots and options were renamed
    map('staticSnapshots.baseUrl', 'static.baseUrl');
    map('staticSnapshots.snapshotFiles', 'static.files');
    map('staticSnapshots.ignoreFiles', 'static.ignore');
    del('staticSnapshots');
  }
}
